lib/hocon/impl/simple_config_origin.rb

```diff
@@ -25,23 +28,173 @@ def initialize(description, line_number, end_line_number,
+             OriginType::FILE,
              url, nil, nil)
   end
 
+  def self.new_url(url)
+    u = url.to_external_form
+    SimpleConfigOrigin.new(u, -1, -1, OriginType::URL, u, nil, nil)
+  end
+
+  def self.new_resource(resource, url = nil)
+    desc = nil
+    if url.nil?
+      desc = resource + " @ " + url.to_external_form
+    else
+      desc = resource
+    end
+    SimpleConfigOrigin.new(desc, -1, -1, OriginType::RESOURCE,
+                           url.nil? ? nil : url.to_external_form,
+                           resource, nil)
+  end
+
+  def with_line_number(line_number)
+    if (line_number == @line_number) and
+        (line_number == @end_line_number)
+      self
+    else
+      Hocon::Impl::SimpleConfigOrigin.new(
+          @_description, line_number, line_number,
+          @origin_type, @url_or_nil, @resource_or_nil, @comments_or_nil)
+    end
+  end
+
+  def add_url(url)
+    SimpleConfigOrigin.new(@_description, line_number, end_line_number, origin_type,
+                           url.nil? ? nil : url.to_external_form, resource_or_nil,
+                           comments_or_nil)
+  end
+
+  def with_comments(comments)
+    if Hocon::Impl::ConfigImplUtil.equals_handling_nil?(comments, @comments_or_nil)
+      self
+    else
+      Hocon::Impl::SimpleConfigOrigin.new(
+          @_description, @line_number, @end_line_number,
+          @origin_type, @url_or_nil, @resource_or_nil, comments)
+    end
+  end
+
+  def prepend_comments(comments)
+    if Hocon::Impl::ConfigImplUtil.equals_handling_nil?(comments, @comments_or_nil)
+      self
+    elsif @comments_or_nil.nil?
+      with_comments(comments)
+    else
+      merged = []
+      merged.concat(comments)
+      merged.concat(@comments_or_nil)
+      with_comments(merged)
+    end
+  end
+
+  def append_comments(comments)
+    if Hocon::Impl::ConfigImplUtil.equals_handling_nil?(comments, @comments_or_nil)
+      self
+    elsif comments_or_nil.nil?
+      with_comments(comments)
+    else
+      merged = comments_or_nil + comments
+      with_comments(merged)
+    end
+  end
+
+  def description
+    if @line_number < 0
+      _description
+    elsif end_line_number == line_number
+      "#{_description}: #{line_number}"
+    else
+      "#{_description}: #{line_number}-#{end_line_number}"
+    end
+  end
+
+  def ==(other)
+    if other.is_a? Hocon::Impl::SimpleConfigOrigin
+      @_description == other._description &&
+          @line_number == other.line_number &&
+          @end_line_number == other.end_line_number &&
+          @origin_type == other.origin_type &&
+          Hocon::Impl::ConfigImplUtil.equals_handling_nil?(@url_or_nil, other.url_or_nil)
+    else
+      false
+    end
+  end
+
+  def hash
+    h = 41 * (41 + @_description.hash)
+    h = 41 * (h + @line_number)
+    h = 41 * (h + @end_line_number)
+    h = 41 * (h + @origin_type.hash)
+
+    unless @url_or_nil.nil?
+      h = 41 * (h + @url_or_nil.hash)
+    end
```

Comment:
I think we need to take `resourceOrNil` into account when computing the hash code